### ISSUE TITLE
Profile cache level

### ISSUE THREAD

**Issue by vitalyzhakov** [yiisoft/yii2#12775]
### What steps will reproduce the problem?
Process connect to several cache systems. For performance purposes programmers need to know how effectively use cache.

### What is the expected result?
In debug panel: performance, time to do query.

### What do you get instead?
Nothing in debug panel if I use memcache.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.9
| PHP version      | 7.0.12
| Operating system | Debian 8, Docker

**Comment by vitalyzhakov** [comment]
@samdark , why moved to debug? I think, Yii::beginProfile must be in core files (for example, in
https://github.com/yiisoft/yii2/blob/master/framework/caching/Cache.php) such as https://github.com/yiisoft/yii2/blob/master/framework/db/Command.php#L842

**Comment by samdark** [comment]
Probably you're right...

**Comment by dynasource** [comment]
this is a useful indeed.

The question is how to implement this in an optimal & flexible way.
Perhabs we want this to be covered with Events inside the Cache.php. Then developers have more flexibility.

Looking at your current PR https://github.com/yiisoft/yii2/issues/12891, I see that the logging is also triggered on production. Performancewise, this is not preferable. There should at least be a check in place for YII_DEBUG.

If you'd ask me, I would go for the triggering of an Event. Then you can decide in the applications bootstrap.php which Events you want to trigger for your debug-toolbar.

**Comment by vitalyzhakov** [comment]
Can't understand your idea, please, give me an example of using events.
Checking YII_DEBUG I will add, thank you.

**Comment by dynasource** [comment]
of course:

in bootstrap.php

```php
use yii\base\Event;
use yii\base\View;

Event::on(View::class,View::EVENT_BEFORE_RENDER,function(){
    Yii::beginProfile(Yii::$app->params['RENDER'] ?? Yii::$app->params['RENDER'] = 0,'View::render()');
    Yii::$app->params['RENDER']++;
});
Event::on(View::class,View::EVENT_AFTER_RENDER,function(){
    Yii::endProfile(Yii::$app->params['RENDER']-1);
});
```

**Comment by samdark** [comment]
Triggering events should have even more performance impact than calling profiling methods so I'd keep it as profiling ones only and in debug mode only.

**Comment by dynasource** [comment]
I thought about that, but it can be questioned:
Firstly, is triggering and empty array of events really a performance killer?
Secondly, the total number of cache hits on any page does usually not exceed 5 hits.

Its good to have some benchmarks to be able to properly assess the question about flexibility v.s. performance.

**Comment by samdark** [comment]
> Firstly, is triggering and empty array of events really a performance killer

Yes but you have to measure it. I did that with pre-release 2.0 version.

> Secondly, the total number of cache hits on any page does usually not exceed 5 hits.

Depends on the project very much. I saw projects with 100s of hits.

**Comment by dynasource** [comment]
hmmm...100s is signifant. Then this is no debate anymore. Also, there are alternative solutions thinkable for the debug environment, such as class replacements with Yii::$container->set() with debugging stepping points.

**Comment by vitalyzhakov** [comment]
I have updated the code https://github.com/yiisoft/yii2/pull/12911/files, please, review it.

**Comment by kidol** [comment]
In other cases profiling doesn't depend on `YII_DEBUG`. If performance really is a problem, I'd rather see `$enableProfiling` property which defaults to `false`.

**Comment by samdark** [comment]
I'd rather see it defaulted to YII_DEBUG... but otherwise a good idea.
